packages/labs/ssr/src/lib/dom-shim.ts

```diff
@@ -109,6 +109,14 @@ export const getWindow = ({
         observedAttributes:
           (ctor as CustomHTMLElement).observedAttributes ?? [],
       });
+      function localName() {
+        return name;
+      }
+      Object.defineProperty(ctor.prototype, 'localName', {
```

Comment:
Rather than add a property to the element, which could conceivably collide with a class property and wouldn't work with scoped registries, and may have problems with inheritance (ie https://github.com/lit/lit/issues/1996#issuecomment-875811141)

I'd rather see a method added the registry to retrieve the tag name. Since this seems somewhat likely to be added in standards in the future, it should have a definitely prollyfill-type name.

Reply:
So do you mean to add `localName` to the HTMLElement shim and then have it call out to the custom element registry? I could do that I guess, but I don't think it would work with scoped registries either.

Reply:
No, I mean add a method to CustomElementRegistry. Like:

```
const tagName = customElements.prollyfillGetTagName(MyElement);
```

The registry can store the names in a map.

Reply:
I would want to also add a getter on HTMLElement for localName, is that ok? The goal for me as an end-user is to only use standard APIs. Ideally I could use Lit with another shim (as I'm doing elsewhere), and that shim won't have `prollyfillGetTagName` but will have localName. 

So adding a `prollyfillGetTagName` and then using that within HTMLElement.prototype.localName makes sense to me.

Reply:
`localName` already exists on the prototype, but for instances and throws if you try to access it directly off the prototype object. Not throwing and returning a value from `MyElement.prototype.localName` is just as non-standard as a patched method on `CustomElementRegistry`. It's also not compatible with scoped registries where different instances with the same prototype might return different tag names.

This functionality could be written as a patch on `CustomElementRegistry.prototype.define` and a library function that takes and registry and constructor to get the name, and that would work on any standard implementation of CustomElementRegistry, even client-side. Would that suit your needs? What I like about that is that it's obvious that this is non-standard behavior from importing the library and the library can document that it needs to be patched in before elements are registered.

Reply:
There might be some confusion here. `localName` does not exist on the prototype in this shim. I am *not* wanting to access it off the prototype. I only want to access it on the instance through `new MyElement().localName`. I can even make it throw.

To be clear, all I want to do here is implement the standard `localName` behavior in this shim.

My first approach was to do it this way, but as you pointed out it's not correct in a number of ways. The shim is not correct in many other ways, so I didn't think correctness was a big concern of the shim. But I'm willing to make this as correct as you want it to be.  But I don't want to use a 3rd party package to access the localName as I don't need that in the browser or anywhere else that supports the Element interface.

Reply:
I think the meta point on this issue is that while it's trivial to add `localName` to the shim (and we could do that), the problem is that even if the shim had `localName`, newing up a CE _just_ to get its `localName` is too wasteful of a pattern to use for mapping a CE constructor to its tagName.

While that's maybe the only standard way to do it, it's just a bad pattern overall and likely worth patching customElements.define yourself to get that information while we wait on a better spec-based way to do it.

Reply:
I'm not sure I agree, creating an object itself is not expensive. It's possible that a constructor might do something expensive but it's generally advised against; most custom element constructors create a shadowRoot and do little else.

Patching customElements.define relies on script ordering, so the user has to run code in a particular order or otherwise they might define the element before the patch has been right, in which case you won't have your map.